code/game/machinery/recharger.dm

```diff
@@ -34,142 +34,153 @@
-		return
-
-	var/allowed = is_type_in_typecache(G, allowed_devices)
-
-	if(allowed)
-		if(anchored)
-			if(charging || panel_open)
-				return 1
-
-			//Checks to make sure he's not in space doing it, and that the area got proper power.
-			var/area/a = get_area(src)
-			if(!isarea(a) || a.power_equip == 0)
-				to_chat(user, span_notice("[src] blinks red as you try to insert [G]."))
-				return 1
-
-			if (istype(G, /obj/item/gun/energy))
-				var/obj/item/gun/energy/E = G
-				if(!E.can_charge)
-					to_chat(user, span_notice("Your gun has no external power connector."))
-					return 1
-
-			if(!user.transferItemToLoc(G, src))
-				return 1
-			setCharging(G)
-
-		else
-			to_chat(user, span_notice("[src] isn't connected to anything!"))
-		return 1
-
-	if(anchored && !charging)
-		if(default_deconstruction_screwdriver(user, "recharger", "recharger", G))
-			update_appearance()
-			return
+/obj/machinery/recharger/attackby(obj/item/attacking_item, mob/user, params)
+	if(!is_type_in_typecache(attacking_item, allowed_devices))
+		return ..()
+
+	. = TRUE
+	if(!anchored)
+		to_chat(user, span_notice("[src] isn't connected to anything!"))
+		return
+	if(charging || panel_open)
+		return
+
+	var/area/our_area = get_area(src) //Check to make sure user's not in space doing it, and that the area got proper power.
+	if(!isarea(our_area) || our_area.power_equip == 0)
+		to_chat(user, span_notice("[src] blinks red as you try to insert [attacking_item]."))
+		return
 
-		if(panel_open && G.tool_behaviour == TOOL_CROWBAR)
-			default_deconstruction_crowbar(G)
+	if (istype(attacking_item, /obj/item/gun/energy))
+		var/obj/item/gun/energy/energy_gun = attacking_item
+		if(!energy_gun.can_charge)
+			to_chat(user, span_notice("Your gun has no external power connector."))
 			return
+	user.transferItemToLoc(attacking_item, src)
 
-	return ..()
+/obj/machinery/recharger/wrench_act(mob/living/user, obj/item/tool)
+	if(charging)
+		to_chat(user, span_notice("Remove the charging item first!"))
+		return TOOL_ACT_SIGNAL_BLOCKING
+	set_anchored(!anchored)
+	power_change()
+	to_chat(user, span_notice("You [anchored ? "attached" : "detached"] [src]."))
+	tool.play_tool_sound(src)
+	return TOOL_ACT_TOOLTYPE_SUCCESS
+
+/obj/machinery/recharger/screwdriver_act(mob/living/user, obj/item/tool)
+	if(!anchored || charging)
+		return TOOL_ACT_SIGNAL_BLOCKING
+	. = default_deconstruction_screwdriver(user, base_icon_state, base_icon_state, tool)
+	if(.)
+		update_appearance()
+
+/obj/machinery/recharger/crowbar_act(mob/living/user, obj/item/tool)
+	return (!anchored || charging) ? TOOL_ACT_SIGNAL_BLOCKING : default_deconstruction_crowbar(tool)
 
 /obj/machinery/recharger/attack_hand(mob/user, list/modifiers)
 	. = ..()
 	if(.)
 		return
 
 	add_fingerprint(user)
-	if(charging)
-		charging.update_appearance()
-		charging.forceMove(drop_location())
-		user.put_in_hands(charging)
-		setCharging(null)
-
+	if(!charging)
+		return
+	if(user.put_in_hands(charging))
+		return
```

Comment:
```suggestion
	if(isnull(charging) || user.put_in_hands(charging))
		return
```

isnull is much more explicit to read in this instance, and about 0.00001 times faster (but fastness isn't the point, it's the explicity goodness)